Quotient/room.cpp

```diff
@@ -1752,12 +1759,51 @@ Room::Private::moveEventsToTimeline(RoomEventsRange events,
         if (auto n = q->checkForNotifications(ti); n.type != Notification::None)
             notifications.insert(eId, n);
         Q_ASSERT(q->findInTimeline(eId)->event()->id() == eId);
+        updateThread(ti.event());
     }
     const auto insertedSize = (index - baseIndex) * placement;
     Q_ASSERT(insertedSize == int(events.size()));
     return Timeline::size_type(insertedSize);
 }
 
+void Room::Private::updateThread(const RoomEvent* event)
+{
+    const auto rme = eventCast<const RoomMessageEvent>(event);
+    if (rme == nullptr) {
+        return;
+    }
+    if (!rme->isThreaded()) {
+        return;
+    }
+
+    if (threads.exists(rme->threadRootEventId())) {
+        auto thread = threads.getThread(rme->threadRootEventId());
+        const auto threadLatestIndex = eventsIndex.constFind(thread->latestEventId());
+        if (threadLatestIndex == eventsIndex.cend()) {
+            // Assume this is the latest event. This shouldn't happen but we can work around it.
+            thread->addEvent(rme, true, rme->senderId() == connection->userId());
+            return;
+        }
+
+        const auto eventIndexIt = eventsIndex.constFind(rme->id());
+        if (eventIndexIt == eventsIndex.cend()) {
+            qCCritical(EVENTS) << rme->id() << "not in the timeline. Update a thread after moving the event to timeline.";
+        }
+
+        thread->addEvent(rme, *eventIndexIt > *threadLatestIndex, rme->senderId() == connection->userId());
+    } else {
+        if (!event->id().isEmpty()) {
+            threads.add(std::make_unique<Thread>(rme->threadRootEventId(),
+                        // For pending events we can get the full correct details when the remote echo comes in.
+                        rme->id().isEmpty() ? rme->threadRootEventId() : rme->id(),
```

Comment:
`rme->id()` is exactly the same as `event->id()` in the enclosing `if`. You either have to remove the enclosing `if` (which is perfectly fine, I think?), or you don't need to special-case pending events here because they won't reach this code.

Reply:
So I removed the special casing for pending events